### PR TITLE
Warn about MRI 1.9 `super` infinite recursion bug.

### LINKED ISSUE #814: stack level too deep (SystemStackError) with rspec-expectations
Hi @myronmarston,

I'm on my way down the rabbit hole to support 1.9.3. and 1.8.7 for aruba. Now I'm facing this error every time I use a negated alias for a matcher on ruby 1.9.3. I use rspec-expectation 3.3.

```
/home/user/.rvm/gems/ruby-1.9.3-p551/gems/rspec-expectations-3.3.0/lib/rspec/matchers.rb:972: stack level too deep (SystemStackError)
```

and I'm not sure if it's a bug or a mistake by me. This is a minimum failing example which works fine on 1.8.7 and > 2.0, but not on 1.9.3. I think rspec's code takes a "different route" for 1.8.7, so maybe it's the way ruby 1.9.3 behaves in contrast to > 2.0.

```ruby
require 'rspec/expectations'

RSpec::Matchers.define :run_too_long do
  match do |actual|
    true
  end
end

RSpec::Matchers.define_negated_matcher :have_finished_in_time, :run_too_long

include RSpec::Matchers

expect('command').to run_too_long
$stderr.puts 'run too long works'

expect('command').to have_finished_in_time
$stderr.puts 'have finished in time not'
```

These are the installed gems:

```
Gems included by the bundle:
  * archive-tar-minitar (0.5.2)
  * aruba (0.8.0)
  * ast (2.0.0)
  * astrolabe (1.3.0)
  * bcat (0.6.2)
  * binding_of_caller (0.7.2)
  * builder (3.2.2)
  * bundler (1.10.5)
  * childprocess (0.5.6)
  * coderay (1.1.0)
  * columnize (0.9.0)
  * contracts (0.9)
  * cucumber (1.3.20)
  * cucumber-pro (0.0.17)
  * debug_inspector (0.0.2)
  * debugger (1.6.8)
  * debugger-linecache (1.2.0)
  * debugger-ruby_core_source (1.3.8)
  * diff-lcs (1.2.5)
  * docile (1.1.5)
  * domain_name (0.5.24)
  * eventmachine (1.0.7)
  * faye-websocket (0.9.2)
  * ffi (1.9.10)
  * fuubar (2.0.0)
  * gherkin (2.12.2)
  * http-cookie (1.0.2)
  * httparty (0.13.5)
  * json (1.8.3)
  * kramdown (1.7.0)
  * license_finder (2.0.4)
  * method_source (0.8.2)
  * mime-types (2.6.1)
  * multi_json (1.11.1)
  * multi_test (0.1.2)
  * multi_xml (0.5.5)
  * netrc (0.10.3)
  * parser (2.2.2.6)
  * powerpack (0.1.1)
  * pry (0.10.1)
  * pry-debugger (0.2.3)
  * pry-doc (0.8.0)
  * pry-stack_explorer (0.4.9.2)
  * rack (1.6.4)
  * rainbow (2.0.0)
  * rake (10.4.2)
  * relish (0.7.1)
  * rest-client (1.8.0)
  * rspec (3.3.0)
  * rspec-core (3.3.1)
  * rspec-expectations (3.3.0)
  * rspec-mocks (3.3.1)
  * rspec-support (3.3.0)
  * rubocop (0.32.1)
  * ruby-progressbar (1.7.5)
  * simplecov (0.10.0)
  * simplecov-html (0.10.0)
  * slop (3.6.0)
  * thor (0.19.1)
  * unf (0.1.4)
  * unf_ext (0.0.7.1)
  * websocket-driver (0.5.4)
  * websocket-extensions (0.1.2)
  * xml-simple (1.1.5)
  * yard (0.8.7.6)
```

This is the installed ruby:
```
ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-linux]
```

Can you have a look at this? Thanks in advance.

Cheers,
Dennis

### PR DESCRIPTION
**Pull request by myronmarston**
Closes #814.

### CODE REVIEW

**Comment by JonRowe** [comment]
Seems there are some specs that generate the warning missing an `allow_warning`

File: lib/rspec/matchers.rb [hunk 1]
```diff
@@ -1005,5 +1005,31 @@ def self.is_a_matcher?(obj)
...
+      # in order to find the cases where it was already previously included.
+      def self.append_features(mod)
+        subclasses = ObjectSpace.each_object(Class).select { |c| c < mod && c < self }
+        return super unless subclasses.any?
+
+        subclasses.reject! { |s| subclasses.any? { |s2| s < s2 } } # Filter to the root ancestor.
+        subclasses = subclasses.map { |s| "`#{s}`" }.join(", ")
+
+        RSpec.warning "`#{self}` has been included in a superclass (`#{mod}`) " \
+                      "after previously being included in subclasses (#{subclasses}), " \
```
Comment (JonRowe) [inline lib/rspec/matchers.rb]: As the travis build is demonstrating this list could be really long...

**Comment by myronmarston** [comment]
> Seems there are some specs that generate the warning missing an allow_warning

There failures are in rspec-core's specs. This is surprising.  I'm going to dig into it now.

File: lib/rspec/matchers.rb (hunk 1 above)
Comment (myronmarston) [inline lib/rspec/matchers.rb]: That was a bug, actually, related to the fact that I didn't handle re-inclusion properly.

**Comment by myronmarston** [comment]
> Seems there are some specs that generate the warning missing an allow_warning

Nope, it was a bug related to re-inclusion that I have now addressed.

**Comment by myronmarston** [comment]
And now it's green :).

**Comment by JonRowe** [comment]
LGTM

**Pull request merged**